### PR TITLE
Vehicle Handling

### PR DESCRIPTION
**Pull request by aaron02**
**Description**
Vehicles should now work as intended. 
Fix Accessory not attaching correctly.
Fix Vehicle Movement
Fix Incorrect Seat Orientations
Fix Vehicles in Cataclysm
Close:
#884 

- [x] Wotlk working
- [x] Cata working
- [x] Mop build

**Todo / Checklist**
- [x] Target is branch develop.
- [x] Detailed description about this pull request.
- [x] Checked AE-Coding standards.

**Tests Performed:** 
- [x] Build AE.
- [x] Server startup.
- [x] Log into world.

***Multiversion Ingame Tests Performed:***
- [x] WotLK
- [x] Cata

### CODE REVIEW

**Comment by Zyres** [comment]
The description looks off, can you describe what is actually fixed?
Tested on WotLK AND Cata? If not, test it.

Why do we add TC-Like Utils if we use it exactly once in our world?
Comments/style does not fit our style, where do these comments come from? Why do we repeat what we already read in the functionname?

File: src/world/Objects/Object.cpp [hunk 1]
```diff
@@ -3847,6 +3853,19 @@ void Object::SendMessageToSet(WorldPacket* data, bool /*bToSelf*/, bool /*myteam
     }
 }
 
+void Object::SendMessageToSet(WorldPacket* data, Player const* skipp)
+{
+    if (!IsInWorld())
+        return;
+
+    uint32 myphase = GetPhase();
```
Comment (schnek) [inline src/world/Objects/Object.cpp:3861]: uint32_t

File: src/shared/Util/CMakeLists.txt.bak [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2014-2022 AscEmu Team <http://www.ascemu.org>
```
Comment (schnek) [inline src/shared/Util/CMakeLists.txt.bak:1]: .bak file delete

**Review by schnek — CHANGES_REQUESTED** [review]

**Comment by schnek** [comment]
https://github.com/AscEmu/AscEmu/issues/522 problem still exists

image

**Comment by aaron02** [comment]
try to move spell out of spelldisable
46598

**Comment by schnek** [comment]
wotlk: everything works

have problems with pets:
in some cases the gun disappears
image

can only move to one place
image

sometimes cannon and pet become enemies (visual only)
image
image

if you go into the car, transfer to the cannon, get out of the cannon and go in, you can get into the textures 😁
image
image

**Comment by aaron02** [comment]
i am not able to reproduce your pet bug or the faction bug
i fixed atleast acessory to base seats. but driver to left or right still dont wants to work :/

**Comment by aaron02** [comment]
@schnek can u retest switching seats should work fine now

**Comment by schnek** [comment]
> @schnek can u retest switching seats should work fine now

will check now

**Comment by aaron02** [comment]
@schnek should be working now with cata and switching seats from and leaving vehicle atleast for me.
sadly there is still a bug there. But i call it rdy for this pr^^

**Comment by schnek** [comment]
in file WorldSession.cpp 
line 2902
> WorldPacketHandlers[CMSG_CHANGE_SEATS_ON_CONTROLLED_VEHICLE].handler = &WorldSession::handleChangeSeatsOnControlledVehicle;

this package currently only supports WotLK, but we use it for Cata.


> @schnek should be working now with cata and switching seats from and leaving vehicle atleast for me.
sadly there is still a bug there. But i call it rdy for this pr^^

Yes, it does work, but not always. 😉

also now you can not get out of the transport and continue to move...

**Comment by aaron02** [comment]
All packets sent are correct at leaving vehicle the exact same packets get sent when entering 😜. You can even move and receive movement packets. You just stand still for yourself. I have no clue why it wont move :/

File: src/world/Objects/Object.cpp [hunk 3]
```diff
@@ -4698,7 +4520,215 @@ void MovementInfo::writeMovementInfo(ByteBuffer& data, [[maybe_unused]]uint16_t
...
+        if (element >= MSETransportGuidByte0 && element <= MSETransportGuidByte7)
+        {
+            if (hasTransportData && transport_guid[element - MSETransportGuidByte0])
+                transport_guid[element - MSETransportGuidByte0] ^= data.readUInt8();
+            continue;
+        }
+
+        switch (element)
+        {
+        case MSEFlags:
```
Comment (Appled) [inline src/world/Objects/Object.cpp:4587]: Indent cases in switch

File: src/world/Objects/Object.cpp (hunk 1 above)
Comment (Appled) [reply src/world/Objects/Object.cpp:3861]: uint32_t

File: src/world/Movement/Spline/MoveSplineFlag.h [hunk 4]
```diff
@@ -112,6 +114,120 @@ class MoveSplineFlag
...
+        // flags that shouldn't be appended into SMSG_MONSTER_MOVE\SMSG_MONSTER_MOVE_TRANSPORT packet, should be more probably
+        Mask_No_Monster_Move = Mask_Final_Facing | Mask_Animations | Done,
+        // CatmullRom interpolation mode used
+        Mask_CatmullRom = Flying | Catmullrom,
+        // Unused, not suported flags
+        Mask_Unused = No_Spline | Enter_Cycle | Frozen | Unknown0 | Unknown2 | Unknown3 | Unknown4 | Unknown6 | Unknown7 | Unknown8 | Unknown9
+    };
+
+    inline uint32_t& raw() { return (uint32_t&)*this; }
+    inline uint32 const& raw() const { return (uint32_t const&)*this; }
```
Comment (Appled) [inline src/world/Movement/Spline/MoveSplineFlag.h:165]: uint32_t

File: src/world/Objects/Object.cpp [hunk 5]
```diff
@@ -4744,143 +4774,149 @@ void MovementInfo::writeMovementInfo(ByteBuffer& data, [[maybe_unused]]uint16_t
...
-            case MSEMovementCounter:
-                data << uint32_t(0);
-                break;
-            case MSECustomSpeed:
-                data << float(custom_speed);
-                break;
-            default:
-                sLogger.failure("Wrong movement status element");
-                break;
+        case MSEHasMovementFlags:
```
Comment (Appled) [inline src/world/Objects/Object.cpp:4777]: Indent cases in switch

File: src/world/Objects/Units/Unit.cpp [hunk 6]
```diff
@@ -7185,3 +7156,364 @@ bool Unit::isLootable()
...
+    return nullptr;
+}
+
+void Unit::handleSpellClick(Unit* clicker, int8_t seatId /*= -1*/)
+{
+    bool spellClickHandled = false;
+    uint32_t spellClickEntry = getVehicleKit() ? getVehicleKit()->getEntry() : getEntry();
+    
+    std::vector<SpellClickInfo> clickBounds = sMySQLStore.getSpellClickInfo(spellClickEntry);
+    for (auto clickPair : clickBounds)
```
Comment (Appled) [inline src/world/Objects/Units/Unit.cpp:7229]: Use 'const auto&' for clickPair or 'auto&' if you need to edit it

File: src/world/Objects/Units/Unit.cpp [hunk 7]
```diff
@@ -7185,3 +7156,364 @@ bool Unit::isLootable()
...
+    {
+        WorldPacket data(SMSG_SPLINE_MOVE_UNROOT, 8);
+        data << GetNewGUID();
+        SendMessageToSet(&data, false);
+    }
+
+    LocationVector pos;
+    // If we ask for a specific exit position, use that one. Otherwise allow scripts to modify it
+    if (exitPosition)
+        pos = *exitPosition;
```
Comment (Appled) [inline src/world/Objects/Units/Unit.cpp:7437]: Use braces for 'if' statement if you used them for 'else' statement

File: src/world/Objects/Units/Unit.cpp [hunk 8]
```diff
@@ -7185,3 +7156,364 @@ bool Unit::isLootable()
...
+    }
+}
+#else
+void Unit::handleSpellClick(Unit* clicker)
+{
+    bool spellClickHandled = false;
+    uint32_t spellClickEntry = getEntry();
+
+    std::vector<SpellClickInfo> clickBounds = sMySQLStore.getSpellClickInfo(spellClickEntry);
+    for (auto clickPair : clickBounds)
```
Comment (Appled) [inline src/world/Objects/Units/Unit.cpp:7492]: Use 'const auto&' for clickPair or 'auto&' if you need to edit it

File: src/world/Server/Packets/Handlers/VehicleHandler.cpp [hunk 9]
```diff
@@ -16,119 +16,187 @@ This file is released under the MIT license. See README-MIT for more information
...
+    vehicle_base->obj_movement_info = movementInfo;
+
+    ObjectGuid accessory = extra.Data.guid;
+    int8_t seatId = extra.Data.byteData;
+
+    if (vehicle_base->getGuid() != movementInfo.guid)
+        return;
+
+    if (!accessory)
+        GetPlayer()->callChangeSeat(seatId, seatId > 0); // prev/next
```
Comment (Appled) [inline src/world/Server/Packets/Handlers/VehicleHandler.cpp:176]: Use braces in 'if' statement if you used them in 'else' statement

File: src/world/Server/Packets/Handlers/VehicleHandler.cpp [hunk 10]
```diff
@@ -138,18 +206,28 @@ void WorldSession::handleRemoveVehiclePassenger(WorldPacket& recvPacket)
         return;
 
     const auto passengerUnit = _player->GetMapMgr()->GetUnit(srlPacket.guid);
-    if (passengerUnit == nullptr)
+    if(!passengerUnit)
```
Comment (Appled) [inline src/world/Server/Packets/Handlers/VehicleHandler.cpp:209]: if (...)

File: src/world/Server/Packets/Handlers/VehicleHandler.cpp [hunk 11]
```diff
@@ -138,18 +206,28 @@ void WorldSession::handleRemoveVehiclePassenger(WorldPacket& recvPacket)
...
+        return;
+
+    if (!passengerUnit->isOnVehicle(vehicle->getBase()))
         return;
 
-    vehicle->EjectPassenger(passengerUnit);
+    auto seat = vehicle->getSeatForPassenger(passengerUnit);
+    if(seat)
```
Comment (Appled) [inline src/world/Server/Packets/Handlers/VehicleHandler.cpp:216]: if (seat)

File: src/world/Storage/MySQLDataStore.hpp [hunk 12]
```diff
@@ -67,7 +67,7 @@ class SERVER_DECL MySQLDataStore
     typedef std::unordered_map<uint32_t, MySQLStructure::ZoneGuards> ZoneGuardsContainer;
     typedef std::unordered_map<uint32_t, MySQLStructure::Battlemasters> BattleMastersContainer;
     typedef std::vector<MySQLStructure::TotemDisplayIds> TotemDisplayIdContainer;
-    typedef std::unordered_map<uint32_t, SpellClickSpell> SpellClickSpellContainer;
+    typedef std::unordered_map<uint32, SpellClickInfo> SpellClickInfoContainer;
```
Comment (Appled) [inline src/world/Storage/MySQLDataStore.hpp:70]: uint32_t

**Review by Appled — CHANGES_REQUESTED** [review]

**Comment by aaron02** [comment]
should be rdy @Zyres the last problems in cata ( cant move when exit vehicle ) will get fixed later on when more and more packets get added in th evehicle part all is fine.

File: src/world/Server/Packets/Handlers/VehicleHandler.cpp (hunk 9 above)
Comment (aaron02) [reply src/world/Server/Packets/Handlers/VehicleHandler.cpp]: cant find this one

File: src/world/Objects/Units/Creatures/Vehicle.cpp [hunk 13]
```diff
@@ -19,6 +19,50 @@ This file is released under the MIT license. See README-MIT for more information
...
+Vehicle::~Vehicle()
+{
+    if (_status == STATUS_DEACTIVATED)
+    {
+        for (SeatMap::const_iterator itr = Seats.begin(); itr != Seats.end(); ++itr)
+            if (!itr->second.isEmpty())
+                sLogger.failure("Vehicle is not Empty");
+    }
+    else
+        sLogger.failure("Vehicle Accessory Status is not on STATUS_DEACTIVATED");
```
Comment (Appled) [inline src/world/Objects/Units/Creatures/Vehicle.cpp:33]: Use braces for 'else' statement if you used them for 'if' statement

File: src/world/Server/Packets/Handlers/VehicleHandler.cpp [hunk 14]
```diff
@@ -16,119 +16,178 @@ This file is released under the MIT license. See README-MIT for more information
...
+    recvPacket.ReadByteSeq(guid[3]);
+    recvPacket.ReadByteSeq(guid[0]);
+    recvPacket.ReadByteSeq(guid[4]);
+    recvPacket.ReadByteSeq(guid[7]);
+
+    if (vehicle_base->getGuid() != movementInfo.guid)
+        return;
+
+    if (!guid)
+        GetPlayer()->callChangeSeat(-1, seatId > 0); // prev/next
```
Comment (Appled) [inline src/world/Server/Packets/Handlers/VehicleHandler.cpp:173]: Use braces for 'if' statement if you used them in 'else if' statement

**Review by Appled — CHANGES_REQUESTED** [review]

**Comment by schnek** [comment]
please reduce it to one сommit 😉😁

**Review by Appled — APPROVED** [review]

**Pull request merged**